### PR TITLE
Add Hyper-V VM Memory metrics

### PR DESCRIPTION
**Pull request by andbanman**
Add `Win32_PerfRawData_BalancerStats_HyperVDynamicMemoryVM` WMI class to monitor VM memory. The class shows metrics for VMs with and without dynamic memory enabled.

This partially addresses #574 #517

Example output:

```
# HELP windows_hyperv_vm_memory_add_operations                                                                                                                                                                                                                                
# TYPE windows_hyperv_vm_memory_add_operations counter                                                                                                                                                                                                                        
windows_hyperv_vm_memory_add_operations{vm="myvm"} 0                                                                                                                                                                                                                          
# HELP windows_hyperv_vm_memory_added                                                                                                                                                                                                                                         
# TYPE windows_hyperv_vm_memory_added counter                                                                                                                                                                                                                                 
windows_hyperv_vm_memory_added{vm="myvm"} 0                                                                                                                                                                                                                                   
# HELP windows_hyperv_vm_memory_physical_guest_visible 'This counter represents the amount of memory visible in the VM.'                                                                                                                                                      
# TYPE windows_hyperv_vm_memory_physical_guest_visible counter                                                                                                                                                                                                                
windows_hyperv_vm_memory_physical_guest_visible{vm="myvm"} 4000                                                                                                                                                                                                               
# HELP windows_hyperv_vm_memory_physical_memory This counter represents the current amount of memory in the VM.                                                                                                                                                               
# TYPE windows_hyperv_vm_memory_physical_memory counter                                                                                                                                                                                                                       
windows_hyperv_vm_memory_physical_memory{vm="myvm"} 4000                                                                                                                                                                                                                      
# HELP windows_hyperv_vm_memory_pressure_average This counter represents the average pressure in the VM.                                                                                                                                                                      
# TYPE windows_hyperv_vm_memory_pressure_average counter                                                                                                                                                                                                                      
windows_hyperv_vm_memory_pressure_average{vm="myvm"} 161                                                                                                                                                                                                                      
# HELP windows_hyperv_vm_memory_pressure_current This counter represents the current pressure in the VM.                                                                                                                                                                      
# TYPE windows_hyperv_vm_memory_pressure_current counter                                                                                                                                                                                                                      
windows_hyperv_vm_memory_pressure_current{vm="myvm"} 161                                                                                                                                                                                                                      
# HELP windows_hyperv_vm_memory_pressure_maximum This counter represents the maximum pressure band in the VM.                                                                                                                                                                 
# TYPE windows_hyperv_vm_memory_pressure_maximum counter                                                                                                                                                                                                                      
windows_hyperv_vm_memory_pressure_maximum{vm="myvm"} 164                                                                                                                                                                                                                      
# HELP windows_hyperv_vm_memory_pressure_minimum This counter represents the minmum pressure band in the VM.                                                                                                                                                                  
# TYPE windows_hyperv_vm_memory_pressure_minimum counter                                                                                                                                                                                                                      
windows_hyperv_vm_memory_pressure_minimum{vm="myvm"} 159                                                                                                                                                                                                                      
# HELP windows_hyperv_vm_memory_remove_operations                                                                                                                                                                                                                             
# TYPE windows_hyperv_vm_memory_remove_operations counter                                                                                                                                                                                                                     
windows_hyperv_vm_memory_remove_operations{vm="myvm"} 0                                                                                                                                                                                                                       
# HELP windows_hyperv_vm_memory_removed                                                                                                                                                                                                                                       
# TYPE windows_hyperv_vm_memory_removed counter                                                                                                                                                                                                                               
windows_hyperv_vm_memory_removed{vm="myvm"} 0
```

### CODE REVIEW

File: collector/hyperv.go [hunk 1]
```diff
@@ -593,6 +605,69 @@ func NewHyperVCollector() (Collector, error) {
...
 		),
+
+		//
+
+		VMMemoryAddedMemory: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, buildSubsystemName("vm_memory"), "added"),
+			"",
+			[]string{"vm"},
+			nil,
+		),
```
Comment (breed808) [inline collector/hyperv.go:616]: Each of the `Counter` metrics will need a `_total` suffix added to the description.

File: collector/hyperv.go [hunk 2]
```diff
@@ -593,6 +605,69 @@ func NewHyperVCollector() (Collector, error) {
...
+		VMMemoryAveragePressure: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, buildSubsystemName("vm_memory"), "pressure_average"),
+			"This counter represents the average pressure in the VM.",
+			[]string{"vm"},
+			nil,
+		),
```
Comment (breed808) [inline collector/hyperv.go:622]: For the `Gauge` metrics, it would be best to remove "counter" from the description to avoid confusion of the metric type.

**Review by breed808 — CHANGES_REQUESTED** [review]
Thanks for submitting this! There's some changes required for the metric names & descriptions, but overall this is looking good :+1:

**Comment by breed808** [comment]
Aside from the review comments, you'll also need to sign off on the commit (see the link next to the failed action for info).

It'd be a good idea to rebase your feature branch to pick up recent CI fixes too.

File: collector/hyperv.go (hunk 2 above)
Comment (andbanman) [reply collector/hyperv.go]: Sounds good. The descriptions started from the WMI property description. Fixed in the latest version.

**Review by breed808 — APPROVED** [review]
Thanks, this is looking good. I've added initial documentation entries for the memory metrics as well.

**Comment by andbanman** [comment]
Thank you for taking the time to review and for the helpful suggestions. I'm working on relevant alerts and a Grafana dashboard that I plan to share soon.

**Pull request merged**